content/chapters/compute/lecture/slides.mdpp

```diff
@@ -0,0 +1,1118 @@
+    * Fast when lock is (almost) immediately acquired
+* Disadvantage:
+    * Overhead: runs on the CPU while waiting for the lock (doing nothing)
+
+---
+
+### Mutual Exclusion - Mutex
+
+```c [1 - 3 | 5 - 9 | 12 - 19 | 21 - 26]
+struct mutex {
+	/* Use internal lock for `atomic_flag_test_and_set`. */
+	int lock;
+
+	/**
+	 * Store WAITING threads in this queue. Move one thread to READY when
+	 * another one calls `mutex_unlock`.
+	 */
+	struct queue *queue;
+};
+
+void mutex_lock(struct mutex *m)
+{
+	while (atomic_flag_test_and_set(&m->lock)) {
+        thread = current_thread();
+		queue_push(m->queue, thread);
+		move_to_waiting(thread);
+    }
+}
+
+void mutex_unlock(struct mutex *m)
+{
+	m->lock = 0;
+	thread = queue_pop(m->queue);
+	move_to_ready(thread);
+}
+```
+
+----
+
+### Mutex - Analysis
+
+* Prevents overhead from busy-waiting
+* Threads move to WAITING state while waiting for the lock
+* Introduces overhead from context switch
+* Releasing a mutex held by another thread is [**undefinded behaviour**](https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_mutex_lock.html)
+
+----
+
+## Who Should Release the Lock?
+
+![Locks Undefined Behaviour](./media/lock-undefined-behaviour.svg)
+
+---
+
+### Mutual Exclusion - Overhead
+
+* The more complex the synchronization mechanism, the higher the overhead:
+
+``` [1 - 2 | 3 - 4 | 5 - 6 | 7 - 8]
+student@os:~/.../compute/lecture/demo/race-condition$ ./race_condition
+var = 1232611; time = 14 ms
+student@os:~/.../compute/lecture/demo/race-condition$ ./race_condition_atomic
+var = 2000000; time = 31 ms
+student@os:~/.../compute/lecture/demo/race-condition$ ./race_condition_spinlock
+var = 2000000; time = 73 ms
+student@os:~/.../compute/lecture/demo/race-condition$ ./race_condition_mutex
+var = 2000000; time = 98 ms
+```
+
+* Why is the spinlock faster?
+    * **Short** critical section, so the lock is acquired fast
+    * The context switches performed by the mutex are costlier
+
+----
+
+## Critical Section - Granularity
+
+* What if we place all the code in `increase_var` in a critical section?
+* Short critical sections mean more frequent context switches
+* Large critical sections mean less parallelism (code runs almost sequentially)
+
+```
+student@os:~/.../compute/lecture/demo/granularity$ ./coarse_granularity
+var = 2000000; time = 36 ms
+student@os:~/.../compute/lecture/demo/granularity$ ./fine_granularity
+var = 2000000; time = 1268 ms
+```
+
+---
+
+## Locks vs Notifications
+
+![Lock vs Notification](./media/lock-vs-notify.svg)
+
+----
+
+## Locks vs Notifications
+
+* Locks:
+    * isolate **critical sections**
+    * thread that calls `unlock()` must hold the lock
+
+* Notifications (conditions and semaphores):
+    * one thread waits for an **event**
+    * **another thread** notifies waiting thread on event completion
+
+----
+
+### Semaphore
+
+* Generalized mutex (mutex = binary semaphore)
+* Lock can take values greater than 1
+* `acquire_lock() ~ lock--`:
```

Comment:
Don't use `lock--` / `lock++`. Use `count`. A semaphore is basically a counter with atomic `increment_and_signal_if_waiters()` and `decrement_and_wait_for_signalers()` primitives.